core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java

```diff
@@ -0,0 +1,388 @@
+          return sql.executeUpdate();
+        }
+      });
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Failed to drop %s", identifier);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted in call to dropTable", e);
+    }
+
+    if (deletedRecords > 0) {
+      LOG.info("Successfully dropped table {}.", identifier);
+    } else {
+      LOG.info("Cannot drop table: {}! table not found in the catalog.", identifier);
+      return false;
+    }
+
+    TableOperations ops = newTableOps(identifier);
+    TableMetadata lastMetadata = ops.current();
+
+    if (purge && lastMetadata != null) {
+      CatalogUtil.dropTableData(ops.io(), lastMetadata);
+      LOG.info("Table {} data purged!", identifier);
+    }
+
+    return true;
+
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    if (!namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    try {
+      return connections.run(conn -> {
+        List<TableIdentifier> results = Lists.newArrayList();
+        try (PreparedStatement sql = conn.prepareStatement(JdbcUtil.LIST_TABLES_SQL)) {
+          sql.setString(1, catalogName);
+          sql.setString(2, JdbcUtil.namespaceToString(namespace));
+
+          ResultSet rs = sql.executeQuery();
+          while (rs.next()) {
+            results.add(JdbcUtil.stringToTableIdentifier(rs.getString(JdbcUtil.TABLE_NAMESPACE), rs.getString(
+                JdbcUtil.TABLE_NAME)));
+          }
+
+          return results;
+        }
+      });
+
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Failed to list tables in namespace: %s", namespace);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted during JDBC operation", e);
+    }
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    try {
+      int updatedRecords = connections.run(conn -> {
+        try (PreparedStatement sql = conn.prepareStatement(JdbcUtil.RENAME_TABLE_SQL)) {
+          // SET
+          sql.setString(1, JdbcUtil.namespaceToString(to.namespace()));
+          sql.setString(2, to.name());
+          // WHERE
+          sql.setString(3, catalogName);
+          sql.setString(4, JdbcUtil.namespaceToString(from.namespace()));
+          sql.setString(5, from.name());
+          return sql.executeUpdate();
+        }
+      });
+
+      if (updatedRecords == 1) {
+        LOG.debug("Successfully renamed table from {} to {}!", from, to);
+      } else if (updatedRecords == 0) {
+        throw new NoSuchTableException("Failed to rename table! Table '%s' not found in the catalog!", from);
+      } else {
+        throw new RuntimeException("Failed to rename table! Rename operation Failed");
```

Comment:
This violates our assumption about tables, but it doesn't mean the operation failed. Javadoc states that this will return the number of affected rows or 0. So if this isn't 0 or 1, then it means that more than 1 row was affected and our underlying primary key assumption is violated. I think in that case we want an error message that highlights the problem at WARNING: "Rename operation affected %s rows: the catalog table's primary key assumption has been violated"